cpp/cycles_module/algorithm/cycles.cpp

```diff
@@ -0,0 +1,208 @@
+
+#include "cycles.hpp"
+
+namespace cycles_util {
+
+NodeState::NodeState(std::size_t number_of_nodes) {
+  visited.resize(number_of_nodes, false);
+  parent.resize(number_of_nodes, std::nullopt);
+  depth.resize(number_of_nodes, 0);
+}
+
+void NodeState::SetVisited(std::uint64_t node_id) { visited[node_id] = true; }
+
+bool NodeState::IsVisited(std::uint64_t node_id) const { return visited[node_id]; }
+
+void NodeState::SetParent(std::uint64_t parent_id, std::uint64_t node_id) { parent[parent_id] = node_id; }
+
+std::optional<std::uint64_t> NodeState::GetParent(std::uint64_t node_id) const { return parent[node_id]; }
+
+void NodeState::SetDepth(std::uint64_t node_id, std::uint64_t node_depth) { depth[node_id] = node_depth; }
+
+std::uint64_t NodeState::GetDepth(std::uint64_t node_id) const { return depth[node_id]; }
+
+void FindNonSpanningTreeEdges(std::uint64_t node_id, const mg_graph::GraphView<> &graph, NodeState *state,
+                              std::set<std::pair<uint64_t, uint64_t>> *non_st_edges) {
+  std::unordered_set<std::uint64_t> unique_neighbour;
+  state->SetVisited(node_id);
+  for (const auto &neigh : graph.Neighbours(node_id)) {
+    auto next_id = neigh.node_id;
+
+    // Check if is returning edge or already visited neighbour
+    if (state->GetParent(node_id) != std::nullopt &&
+        (next_id == state->GetParent(node_id) || unique_neighbour.find(next_id) != unique_neighbour.end())) {
+      continue;
+    }
+
+    unique_neighbour.insert(next_id);
+
+    if (state->IsVisited(next_id)) {
+      auto sorted_edge = std::minmax(node_id, next_id);
+      non_st_edges->insert(sorted_edge);
+      continue;
+    }
+
+    // Set depth and parent for the next ST iteration
+    state->SetParent(next_id, node_id);
+    state->SetDepth(next_id, state->GetDepth(node_id) + 1);
+    FindNonSpanningTreeEdges(next_id, graph, state, non_st_edges);
+  }
+}
+
+void FindFundamentalCycles(const std::set<std::pair<std::uint64_t, std::uint64_t>> &non_st_edges,
+                           const NodeState &state, std::vector<std::vector<std::uint64_t>> *fundamental_cycles) {
+  for (const auto [from, to] : non_st_edges) {
+    fundamental_cycles->emplace_back(FindFundametalCycle(from, to, state));
+  }
+}
+
+std::vector<std::uint64_t> FindFundametalCycle(std::uint64_t node_a, std::uint64_t node_b, const NodeState &state) {
+  std::vector<std::uint64_t> cycle;
+  if (state.depth[node_a] < state.depth[node_b]) {
+    std::swap(node_a, node_b);
+  }
+
+  // climb until a and b reach the same depth:
+  //
+  //       ()               ()
+  //      /  \             /  \.
+  //    ()   (b)   -->   (a)  (b)
+  //   /  \             /   \.
+  // (a)  ()           ()   ()
+  cycle.emplace_back(node_a);
+  while (state.depth[node_a] > state.depth[node_b]) {
+    if (state.GetParent(node_a) == std::nullopt) {
+      throw mg_exception::InvalidIDException();
+    }
+    node_a = state.GetParent(node_a).value();
+    cycle.emplace_back(node_a);
```

Comment:
optional has really similar usage to the `std::unique_ptr`. You can use * operator to get underlying element or -> operator to call a method on the data saved inside the `std::optional` (if it exists of course).

Here's a typical usage for your use case.
```suggestion
    const auto maybe_parent = state.GetParent(node_a)
    if (!maybe_parent) {
      throw mg_exception::InvalidIDException();
    }
    node_a = *maybe_parent;
    cycle.emplace_back(node_a);
```